[JENKINS-53792] Env vars to configure CLI -auth option

fcojfernandez opened the pull request:
See [JENKINS-53792](https://issues.jenkins-ci.org/browse/JENKINS-53792).

In a similar way that the environment variable _JENKINS_URL_ allows to execute different CLI command without specifying the _-s_ option continuously, environment variable such as _JENKINS_USER_ID_ and _JENKINS_API_TOKEN_ might help configuring easily the executions without specifying the _-auth_ option in each executed command.

**Idea**

Before
```
java -jar jenkins-cli.jar -s https://jenkins.mycompany.com/ -auth my-user:1234567890abcdef my-command
```

After
```
export JENKINS_URL=https://jenkins.mycompany.com/
export JENKINS_USER_ID=my-user
export JENKINS_API_TOKEN=1234567890abcdef
java -jar jenkins-cli.jar my-command
```

In case the _-auth_ option is specified, its value will prevail, although the env vars will not be unset for later use.

### Proposed changelog entries

* JENKINS-53792: Allow JENKINS_USER_ID and JENKINS_API_TOKEN environment variables to configure the authentication when the CLI is invoked.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). 
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs **_N/A_**


### Desired reviewers

@reviewbybees @jenkinsci/code-reviewers

fcojfernandez committed:
[JENKINS-53792] Add env vars for auth option

fcojfernandez committed:
[JENKINS-53792] Fix encoding after first commit

MRamonLeon reviewed (commented):
Some comments about the code. In addition:
I think that writing down the name of the environment variables added in the changelog may be helpful. Therefore I suggest to change the proposed changelog to add those.

varyvol reviewed (approved):
Maybe some code in the tests could be refactored out (especially invoking the command or creating the maps with the env variables), but looks good!

fcojfernandez committed:
[JENKINS-53792] Better readability

MRamonLeon commented on cli/pom.xml:
```diff
@@ -77,6 +77,11 @@
+      <groupId>commons-lang</groupId>
+      <artifactId>commons-lang</artifactId>
+      <version>2.6</version>
+    </dependency>
```
Ah, the fine. I thought it was a completely new one

Wadeck replied:
In that case, do we want to put it in the parent pom instead to ease maintenability?

MRamonLeon commented on cli/src/main/resources/hudson/cli/client/Messages.properties:
```diff
@@ -13,10 +13,12 @@ CLI.Usage=Jenkins CLI\n\
 CLI.NoURL=Neither -s nor the JENKINS_URL env var is specified.
 CLI.VersionMismatch=Version mismatch. This CLI cannot work with this Jenkins server.
 CLI.NoSuchFileExists=No such file exists: {0}
+CLI.BadAuth=The JENKINS_USER_ID and JENKINS_API_TOKEN env vars have not been properly set. 
```
As @Wadeck told me, it's not necessary. You can only translate to the languages you know.

Wadeck replied:
That's a good reason to not translate to language you are not familiar with ;)

MRamonLeon commented on test/src/test/java/hudson/cli/CLIEnvVarTest.java:
```diff
@@ -0,0 +1,207 @@
+
+        // Using -auth option
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+            assertEquals(0, launch(Collections.emptyMap(), baos, null,
```
I understand now.

Just in case the speed were a concern:
You can change it with: https://dzone.com/articles/how-to-change-environment-variables-in-java

Or you can change your code to get the value from `System.getenv` and if it doesn't exist, from `System.getProperty`. The last one can be set with `System.setProperty` and can be useful with tests.

fcojfernandez replied:
I tried something similar to your proposal, but it only worked if I read the variable in the test method. Reading it from an internal class (the `CLI` class in our case), returned always null, as if the environment variable was not set. And after that, I discovered the `decorateByEnv` method that eased my life!

Regarding the property system, I would prefer to keep the same behaviour than JENKINS_URL has.

MRamonLeon commented on test/src/test/java/hudson/cli/CLIEnvVarTest.java:
```diff
@@ -0,0 +1,207 @@
+
+        // Without -auth option and without env vars
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+            assertEquals(0, launch(Collections.emptyMap(), baos, null,
```
I meant that testing the command without the auth env vars isn't in the scope of the changes. But as you have added new test methods, unrelated to improve the coverage, it sounds good to me.

MRamonLeon reviewed (approved):
I would change the way as invoking the command, but anyway it looks fine enough.

varyvol commented on cli/src/main/java/hudson/cli/CLI.java:
```diff
@@ -563,6 +568,17 @@ public boolean verify(String s, SSLSession sslSession) {
+            // -auth option not set
+            if (StringUtils.isNotBlank(userIdEnv) && StringUtils.isNotBlank(tokenEnv)) {
+                auth = StringUtils.defaultString(userIdEnv).concat(":").concat(StringUtils.defaultString(tokenEnv));
+            } else if (StringUtils.isBlank(userIdEnv) || StringUtils.isBlank(tokenEnv)) {
```
I think this is not right now? If both are blank, the else if would apply and I don't think that's what you want.

Wadeck replied:
Before correcting, please add a test :)

fcojfernandez replied:
@varyvol is right and I shouldn't change the check (Sorry, I did without thinking). Just see my next comment `// Otherwise, none credentials were set`
@Wadeck the test exists (See the `// Without -auth option and without env vars` test) but as I'm using the `who-am-i` command, this error didn't appear. I decided use that CLI command because it allows us to know if the authentication is correct.

Wadeck replied:
> // Without -auth option and without env vars

I meant a test with only one of the two variables, that would have detected the bug :)

Wadeck commented on cli/src/main/resources/hudson/cli/client/Messages_da.properties:
```diff
@@ -26,3 +26,4 @@ Brug: java -jar jenkins-cli.jar [-s URL] command [opts...] args...\n\
 Tilvalg:\n\
 De tilg\u00e6ngelige kommandoer afh\u00e6nger af serveren. K\u00f8r 'help' kommandoen for at se listen.
 CLI.NoURL=Hverken -s eller JENKINS_URL milj\u00f8variablen er defineret
+CLI.BadAuth=JENKINS_USER_ID og JENKINS_API_TOKEN miljovariablerne er ikke korrekt indstillet.
```
I don't know what is the expected process in such case. Google Translate is not known to be accurate but as you say, at least there is something. OTOH sometimes the bad translation is worse than an english wording for non-native people.

Just a -0.1 from my PoV.

Wadeck commented on test/src/test/java/hudson/cli/CLIEnvVarTest.java:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    @Test
+    public void testJENKINS_URL() throws Exception {
```
I will say in addition, that's a perfect move! Adding test coverage especially where there is a lack of test, is always a good point. 👍

Wadeck commented on cli/src/main/java/hudson/cli/CLI.java:
```diff
@@ -452,6 +453,10 @@ public static int _main(String[] _args) throws Exception {
         String auth = null;
+
+        String userIdEnv = System.getenv("JENKINS_USER_ID");
+        String tokenEnv = System.getenv("JENKINS_API_TOKEN");
```
I am not satisfied with the name as it can contain API Token OR password. I don't have a better name at hand but I feel it so confusing.

WDYT to use a single env variable like `JENKINS_CLI_AUTH` and just checking the `<userId>:<tokenOrPassword>` format instead of the double check on their existence ? That could simplify a bit the code and reduce the very hard task to find a better name for the second variable ;)

Wadeck reviewed (commented):
Perhaps some tests are missing:
- using the userInfo from the URL in addition to auth => seems that the userInfo should take precedence. IOW: UserInfo > Auth > env.
- using a real API Token instead of just the password. In the case someone is changing the way the credentials are passed.

More generally perhaps this documentation could help: https://jenkins.io/doc/developer/security/ and in any case it would require an update with this new option.

fcojfernandez commented on cli/src/main/java/hudson/cli/CLI.java:
```diff
@@ -563,6 +568,17 @@ public boolean verify(String s, SSLSession sslSession) {
+            if (StringUtils.isNotBlank(userIdEnv) && StringUtils.isNotBlank(tokenEnv)) {
+                auth = StringUtils.defaultString(userIdEnv).concat(":").concat(StringUtils.defaultString(tokenEnv));
+            } else if (StringUtils.isBlank(userIdEnv) || StringUtils.isBlank(tokenEnv)) {
+                LOGGER.warning(Messages.CLI_BadAuth());
```
Sure thing!

fcojfernandez committed:
[JENKINS-53792] Refactor and use the ApiToken

fcojfernandez commented:
@Wadeck Last commit:
- Use of Api Token instead of password
- Split the test into individuals tests
- Messages only in English and Spanish (Updated this one)

@MRamonLeon I've updated the changelog entry

MRamonLeon commented on cli/src/main/resources/hudson/cli/client/Messages_it.properties:
```diff
@@ -1,22 +1,22 @@
 CLI.NoSuchFileExists=File non esistente: {0}
-CLI.NoURL=Non sono stati specificati n� -s n� la variabile d''ambiente JENKINS_URL.
-CLI.VersionMismatch=Le versioni non corrispondono. Quest''interfaccia della riga di comando non pu� funzionare con questo server Jenkins.
+CLI.NoURL=Non sono stati specificati n� -s n� la variabile d''ambiente JENKINS_URL.
```
is it normal that those characters are shown wrong? Changing then by their unicode codes would avoid troubles. Like in the spanish version.

fcojfernandez committed:
[JENKINS-53792] Fix encoding

oleg-nenashev commented on cli/src/main/java/hudson/cli/CLI.java:
```diff
@@ -563,6 +568,17 @@ public boolean verify(String s, SSLSession sslSession) {
+            } else if (StringUtils.isNotBlank(userIdEnv) || StringUtils.isNotBlank(tokenEnv)) {
+                printUsage(Messages.CLI_BadAuth());
+                return -1;
+            } // Otherwise, none credentials were set
```
... by env vars

oleg-nenashev reviewed (commented):
The pull request looks OK to me as a code.
On the other hand, I am not sure that passing credentials in Environment variables is a good idea from the security PoV. They may be read by another process.

Please put in on hold until it is discussed

oleg-nenashev reviewed (approved):
approved as a code tho

oleg-nenashev reviewed (changes requested):
After some consideration, I request changes in the documentation. Passing credentials as `@FILE` should explicitly recommended as a security BKM. Then this change can be merged IMO

fcojfernandez commented:
@oleg-nenashev I've opened the [jenkins.io#1811](https://github.com/jenkins-infra/jenkins.io/pull/1811)

oleg-nenashev reviewed (changes requested):
Hmm... I would expect cli/src/main/resources/hudson/cli/client/Messages.properties to reference the recommended file option as well (or to refer the jenkins.io doc)

oleg-nenashev commented:
Since I merged the docs already (sorry @daniel-beck), I will get it merged & create a doc patch in a follow-up
